nucypher/policy/conditions/auth.py

```diff
@@ -0,0 +1,99 @@
+from enum import Enum
+from typing import List
+
+import maya
+from eth_account.account import Account
+from eth_account.messages import HexBytes, encode_typed_data
+from siwe import SiweMessage, VerificationError
+
+
+class Auth:
+    class AuthScheme(Enum):
+        EIP712 = "EIP712"
+        SIWE = "SIWE"
+
+        @classmethod
+        def values(cls) -> List[str]:
+            return [scheme.value for scheme in cls]
+
+    class InvalidData(Exception):
+        pass
+
+    class AuthenticationFailed(Exception):
+        pass
+
+    @classmethod
+    def authenticate(cls, data, signature, expected_address):
+        raise NotImplementedError
+
+    @classmethod
+    def from_scheme(cls, scheme: str):
+        if scheme == cls.AuthScheme.EIP712.value:
+            return EIP712Auth
+        elif scheme == cls.AuthScheme.SIWE.value:
+            return SIWEAuth
+
+        raise ValueError(f"Invalid authentication scheme: {scheme}")
+
+
+class EIP712Auth(Auth):
+    @classmethod
+    def authenticate(cls, data, signature, expected_address):
+        try:
+            # convert hex data for byte fields - bytes are expected by underlying library
+            # 1. salt
+            salt = data["domain"]["salt"]
+            data["domain"]["salt"] = HexBytes(salt)
+            # 2. blockHash
+            blockHash = data["message"]["blockHash"]
+            data["message"]["blockHash"] = HexBytes(blockHash)
+
+            signable_message = encode_typed_data(full_message=data)
+            address_for_signature = Account.recover_message(
+                signable_message=signable_message, signature=signature
+            )
+        except Exception as e:
+            # data could not be processed
+            raise cls.InvalidData(
+                f"Invalid EIP712 message: {str(e) or e.__class__.__name__}"
+            )
+
+        if address_for_signature != expected_address:
+            # verification failed - addresses don't match
+            raise cls.AuthenticationFailed(
+                f"EIP712 verification failed; signature does not match expected address, {expected_address}"
+            )
+
+
+class SIWEAuth(Auth):
+    # TODO; this is a safety precaution - is this what we want and is this the correct value?
```

Comment:
ID

Reply:
I understand this as the duration after which the signed authentication message is no longer valid. For an end-user consuming a stream of data, would this mean they would be prompted for a signature again every 2 hours?

Reply:
>I understand this as the duration after which the signed authentication message is no longer valid. For an end-user consuming a stream of data, would this mean they would be prompted for a signature again every 2 hours?

This duration only refers to the time the app will wait since the signature is requested.

I.e. if the user receives on Metamask (or whatever) the signature requests, and the user just leaves the request and after that two-hours-timeout the user does the signature, this signature will not be considered valid.

Reply:
> For an end-user consuming a stream of data, would this mean they would be prompted for a signature again every 2 hours?

If there is a current decryption request that uses a SIWE signature more than 2 hours old, then decryption would fail. Ideally `taco-web`/`taco-auth` would ensure that the signature is refreshed for a decryption operation if the original signature is cutting it close/stale.